### PR TITLE
BFT fixes

### PR DESCRIPTION
**Pull request by andr3ej**
* Fix BFT config that only worked with clusters of size 4
* Overlapping ports check no longer assumes all on localhost
* Fix overlapping ports test to actually check the messages
* Shutdown BFT properly so that tests can run back to back in theory
* Add test for non-trivial BFT cluster, disabled while I investigate why the cluster fails to initialise
* Shutdown NodeBasedTest nodes in parallel, as BFT shutdown time is non-trivial

### CODE REVIEW

File: node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt [hunk 1]
```diff
@@ -1,84 +1,122 @@
...
+            }
+            val spendTxs = (1..10).map {
+                signInitialTransaction(notary, true) {
+                    addInputState(issueTx.tx.outRef<ContractState>(0))
+                }
+            }
+            assertEquals(spendTxs.size, spendTxs.map { it.id }.distinct().size)
+            val results = spendTxs.map { NotaryFlow.Client(it) }.map { services.startFlow(it) }.map { ErrorOr.catch { it.resultFuture.getOrThrow() } }
+            val successfulIndex = results.mapIndexedNotNull { index, result ->
+                if (result.error != null) null else {
```
Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt]: would be a bit more readable with `else` on a new line

File: node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt [hunk 2]
```diff
@@ -1,84 +1,122 @@
...
+        val aliceFuture = startNode(ALICE.name)
+        val notary = bftNotaryCluster(clusterSize).getOrThrow()
+        aliceFuture.getOrThrow().run {
+            val issueTx = signInitialTransaction(notary) {
+                addOutputState(DummyContract.SingleOwnerState(owner = info.legalIdentity))
+            }
+            database.transaction {
+                services.recordTransactions(singleton(issueTx))
+            }
+            val spendTxs = (1..10).map {
```
Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt]: why do we need 10 spend transactions?

File: node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt [hunk 3]
```diff
@@ -1,84 +1,122 @@
...
+                        assertTrue(it in (notary.owningKey as CompositeKey).leafKeys)
+                    }
+                    index
+                }
+            }.single()
+            spendTxs.zip(results).forEach { (tx, result) ->
+                if (result.error != null) {
+                    val error = (result.error as NotaryException).error as NotaryError.Conflict
+                    assertEquals(tx.id, error.txId)
+                    error.conflict.verified().stateHistory.entries.iterator().let {
```
Comment (adagys) [inline node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt]: not sure an iterator is useful here, could just do something like
```
val conflictingStateMap = error.conflict.verified().stateHistory
val (stateRef, consumingTx) = conflictingStateMap.entries.single()
```

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt [hunk 4]
```diff
@@ -85,16 +91,13 @@ object BFTSMaRt {
...
-            val comparator = buildResponseComparator()
-            val extractor = buildExtractor()
-            return ServiceProxy(clientId, configHome.toString(), comparator, extractor)
+            while (true) {
+                val inactive = sessionTable.entries.mapNotNull { if (it.value.channel.isActive) null else it.key }
+                if (inactive.isEmpty()) break
+                log.info("Client-replica channels not yet active: $clientId to $inactive")
+                Thread.sleep((inactive.size * 100).toLong())
+            }
+            return proxy.invokeOrdered(CommitRequest(transaction, otherSide).serialize().bytes).deserialize<ClusterResponse>()
```
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: could you maybe keep the separate vals? slightly easier on the eyes

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt [hunk 5]
```diff
@@ -136,40 +139,53 @@ object BFTSMaRt {
         }
     }
 
+    private class CordaServiceReplica(replicaId: Int, configHome: Path, owner: DefaultRecoverable) : ServiceReplica(replicaId, configHome.toString(), owner, owner, null, DefaultReplier()) {
```
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: maybe add a comment why we need to extend ServiceReplica

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt [hunk 6]
```diff
@@ -136,40 +139,53 @@ object BFTSMaRt {
...
         override fun appExecuteUnordered(command: ByteArray, msgCtx: MessageContext): ByteArray? {
             throw NotImplementedError("No unordered operations supported")
         }
 
         override fun appExecuteBatch(command: Array<ByteArray>, mcs: Array<MessageContext>): Array<ByteArray?> {
-            val replies = command.zip(mcs) { c, _ ->
-                executeCommand(c)
-            }
-            return replies.toTypedArray()
+            return Arrays.stream(command).map(this::executeCommand).toTypedArray()
```
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: I don't think streaming adds any benefit here, as we evaluate it to a list anyway

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt [hunk 7]
```diff
@@ -85,16 +91,13 @@ object BFTSMaRt {
...
-            val response = responseBytes.deserialize<ClusterResponse>()
-            return response
-        }
-
-        private fun buildProxy(configHome: Path): ServiceProxy {
-            val comparator = buildResponseComparator()
-            val extractor = buildExtractor()
-            return ServiceProxy(clientId, configHome.toString(), comparator, extractor)
+            while (true) {
+                val inactive = sessionTable.entries.mapNotNull { if (it.value.channel.isActive) null else it.key }
```
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: what was the issue here?

Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: I'm not sure exactly. The symptom was that the replicas would go around in circles and the cluster would never initialise. I haven't seen that behaviour, locally or in TC, after making this change. The original code created the client lazily (but not lazily enough) which makes me think I'm on the right track here.

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt (hunk 6 above)
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: Where is the list? The map here is Stream -> Stream.

File: node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt [hunk 8]
```diff
@@ -1,84 +1,122 @@
...
+    private fun detectDoubleSpend(faultyReplicas: Int) {
+        val clusterSize = minClusterSize(faultyReplicas)
+        val aliceFuture = startNode(ALICE.name)
+        val notary = bftNotaryCluster(clusterSize).getOrThrow()
+        aliceFuture.getOrThrow().run {
+            val issueTx = signInitialTransaction(notary) {
+                addOutputState(DummyContract.SingleOwnerState(owner = info.legalIdentity))
+            }
+            database.transaction {
+                services.recordTransactions(singleton(issueTx))
```
Comment (andr3ej) [inline node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt]: Done

File: node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt [hunk 9]
```diff
@@ -1,84 +1,122 @@
...
+            }
+            val spendTxs = (1..10).map {
+                signInitialTransaction(notary, true) {
+                    addInputState(issueTx.tx.outRef<ContractState>(0))
+                }
+            }
+            assertEquals(spendTxs.size, spendTxs.map { it.id }.distinct().size)
+            val results = spendTxs.map { NotaryFlow.Client(it) }.map { services.startFlow(it) }.map { ErrorOr.catch { it.resultFuture.getOrThrow() } }
```
Comment (andr3ej) [inline node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt]: Done

File: node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt (hunk 1 above)
Comment (andr3ej) [inline node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt]: Done

File: node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt (hunk 3 above)
Comment (andr3ej) [inline node/src/integration-test/kotlin/net/corda/node/services/BFTNotaryServiceTests.kt]: Thanks, that's much nicer

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt (hunk 4 above)
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: Done

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt (hunk 5 above)
Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: Done

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt (hunk 6 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: Yes but toTypedArray() will evaluate the stream, and in this case it's pretty much the same as simply using a Kotlin `map` on the command array

Comment (andr3ej) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: The existing code creates a list in zip, then copies that list into an array. The new code makes the array directly without the intermediate list.

**Comment by andr3ej** [comment]
@adagys back to you

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt (hunk 6 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: Yeah, I'm just saying that in this particular case this won't have any measurable performance improvements. Anyway, it's not really a problem.

File: node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt (hunk 7 above)
Comment (adagys) [inline node/src/main/kotlin/net/corda/node/services/transactions/BFTSMaRt.kt]: Oh and could you move this loop into a separate function and add a short explanation to the api doc

**Comment by andr3ej** [comment]
note to self: go ahead and merge as soon as the button allows it

**Pull request merged**